### PR TITLE
Add `FrozenTrial.__lt__` to sort trials without key.

### PR DESCRIPTION
**Pull request by toshihikoyanase**
This PR makes `FrozenTrial.__lt__` sortable by `number` to improve backward compatibility with the `FrozenTrial` in `v0.18.x`.

The previous implementation of `FrozenTrial` was based on `NamedTuple` and it was implicitly sorted as a tuple (i.e., the first key is `number` and the second key is `state` and so on). In this PR, I only used `number` because it is unique in a study.

I'm not fully confident with this design of `sort`. Please close PR if the current behavior is expected.

### CODE REVIEW

File: optuna/structs.py [hunk 1]
```diff
@@ -135,6 +135,15 @@ def __ne__(self, other):
 
         return not self.__eq__(other)
 
+    def __lt__(self, other):
+        # type: (Any) -> bool
+
+        if not isinstance(other, type(self)):
+            raise TypeError('\'<\' not supported between instances of {} and {}'.format(
```
Comment (c-bata) [inline optuna/structs.py]: `__lt__()` should return `NotImplemented` constants if it's not supported.
https://docs.python.org/3/library/constants.html#NotImplemented

File: optuna/structs.py [hunk 2]
```diff
@@ -135,6 +135,15 @@ def __ne__(self, other):
 
         return not self.__eq__(other)
 
+    def __lt__(self, other):
+        # type: (Any) -> bool
+
+        if not isinstance(other, type(self)):
```
Comment (c-bata) [inline optuna/structs.py]: [nits] I mind `type(self)` might be the cause of something problem if we will define subclasses. I think `isinstance(other, FrozenTrial)` is more generic.

File: optuna/structs.py (hunk 1 above)
Comment (toshihikoyanase) [reply optuna/structs.py]: Thank you for your suggestion. I've never known the usage of `NotImplemented`.
Fixed: https://github.com/optuna/optuna/pull/719/commits/e7a357802e555079bd9836ed0d1f83404ba17ca8

File: optuna/structs.py (hunk 2 above)
Comment (toshihikoyanase) [reply optuna/structs.py]: That makes sense.
Fixed: https://github.com/optuna/optuna/pull/719/commits/c1ae028258d2cd63f00d335ea57ab458325d5293

File: tests/test_structs.py [hunk 3]
```diff
@@ -116,6 +96,50 @@ def test_frozen_trial_eq_ne():
...
+    assert trial <= trial_other
+    assert not trial_other <= trial
+
+    with pytest.raises(TypeError):
+        trial <= 1
+
+    # A list of FrozenTrials is sortable.
+    trials = [trial_other, trial]
+    trials.sort()
+    assert trials[0] == trial
```
Comment (hvy) [inline tests/test_structs.py]: A nitpick but you can probably check for object identity here.
```suggestion
    assert trials[0] is trial
```

File: tests/test_structs.py [hunk 4]
```diff
@@ -116,6 +96,50 @@ def test_frozen_trial_eq_ne():
...
+    assert trials[1] == trial_other
```
Comment (hvy) [inline tests/test_structs.py]: Ditto.
```suggestion
    assert trials[1] is trial_other
```

**Comment by hvy** [comment]
LGTM, though I left some rather trivial nitpicks in one of the tests.

File: tests/test_structs.py (hunk 3 above)
Comment (toshihikoyanase) [reply tests/test_structs.py]: Thank you for the careful review. You're absolutely right.
Fixed: https://github.com/optuna/optuna/pull/719/commits/0245b32945a696acbe28da7e282070ef07f555b5

**Comment by hvy** [comment]
LGTM!

**Pull request merged**